src/prolong_restrict/pr_ops.hpp

```diff
@@ -371,6 +371,91 @@ struct ProlongateInternalAverage {
   }
 };
 
+struct ProlongateInternalTothAndRoe {
+  static constexpr bool OperationRequired(TopologicalElement fel,
+                                          TopologicalElement cel) {
+    return (cel == TE::CC) && (GetTopologicalType(fel) == TopologicalType::Face);
+  }
+  // Here, fel is the topological element on which the field is defined and
+  // cel is the topological element on which we are filling the internal values
+  // of the field. So, for instance, we could fill the fine cell values of an
+  // x-face field within the volume of a coarse cell. This is assumes that the
+  // values of the fine cells on the elements corresponding with the coarse cell
+  // have been filled.
+  template <int DIM, TopologicalElement fel = TopologicalElement::CC,
+            TopologicalElement cel = TopologicalElement::CC>
+  KOKKOS_FORCEINLINE_FUNCTION static void
+  Do(const int l, const int m, const int n, const int k, const int j, const int i,
+     const IndexRange &ckb, const IndexRange &cjb, const IndexRange &cib,
+     const IndexRange &kb, const IndexRange &jb, const IndexRange &ib,
+     const Coordinates_t &coords, const Coordinates_t &coarse_coords,
+     const ParArrayND<Real, VariableState> *,
+     const ParArrayND<Real, VariableState> *pfine) {
+    using namespace util;
+
+    if constexpr (!IsSubmanifold(fel, cel)) {
```

Comment:
Can you dumb this branch down for me? i.e., what is it checking again (without using the word submanifold)?

Reply:
It is just checking if the topological element `fel` corresponds to some part of the boundary of `cel`, e.g. if `cel = TE::F1` then this would be true for `fel = (TE::NN, TE::E2, TE::E3)` but false for all other elements. Internal prolongation only has to occur on these elements.